### PR TITLE
Adds conversion functions between Asio and Boost.Asio

### PR DESCRIPTION
**Pull request by inetic**

### CODE REVIEW

File: include/maidsafe/common/convert.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#include "asio/ip/address_v4.hpp"
+#include "asio/ip/address_v6.hpp"
+#include "asio/ip/udp.hpp"
+
+#include "boost/asio/ip/address.hpp"
+#include "boost/asio/ip/address_v4.hpp"
+#include "boost/asio/ip/address_v6.hpp"
+#include "boost/asio/ip/udp.hpp"
+
+namespace maidsafe { namespace common { namespace convert {
```
Comment (Fraser999) [inline include/maidsafe/common/convert.h]: We don't usually combine several namespaces on one line like this; it's usually one per line with a blank line in between each.  Probably your best bet is to use clang-format everywhere.

File: include/maidsafe/common/convert.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+// Warning: Will map error from any error_category to an error
+// in the generic_category. Even boost asio category maps to
+// the std generic category so there probably isn't a bijection
+// between the std::error_code and the boost::system::error_code.
+inline std::error_code ToStd(const boost::system::error_code& ec) {
```
Comment (Fraser999) [inline include/maidsafe/common/convert.h]: Shouldn't we `#include <system_error>` and `#include "boost/system/error_code.hpp"` for this?

File: include/maidsafe/common/convert.h (hunk 1 above)
Comment (inetic) [inline include/maidsafe/common/convert.h]: Changed it

File: include/maidsafe/common/convert.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  return asio::ip::address_v6(addr.to_bytes(), addr.scope_id());
+}
+
+inline boost::asio::ip::address ToBoost(const asio::ip::address& addr) {
+  if (addr.is_v4()) {
+    return boost::asio::ip::address(ToBoost(addr.to_v4()));
+  } else if (addr.is_v6()) {
+    return boost::asio::ip::address(ToBoost(addr.to_v6()));
+  } else {
+    assert(0 && "Unknown IP version");
```
Comment (Fraser999) [inline include/maidsafe/common/convert.h]: and `#include <cassert>` for `assert`

File: src/maidsafe/common/tests/convert_test.cc [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    available at: http://www.maidsafe.net/licenses
+
+    Unless required by applicable law or agreed to in writing, the MaidSafe Software distributed
+    under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS
+    OF ANY KIND, either express or implied.
+
+    See the Licences for the specific language governing permissions and limitations relating to
+    use of the MaidSafe Software.                                                                 */
+
+#include "maidsafe/common/log.h"
```
Comment (Fraser999) [inline src/maidsafe/common/tests/convert_test.cc]: I don't think we need to include log.h

File: src/maidsafe/common/tests/convert_test.cc [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+#include "maidsafe/common/test.h"
+#include "maidsafe/common/convert.h"
+
+namespace maidsafe {
+
+namespace test {
+
+struct Conversions : public ::testing::Test {
```
Comment (Fraser999) [inline src/maidsafe/common/tests/convert_test.cc]: This base class isn't really needed.  Can't we just use gtest's `TEST` macro rather than `TEST_F` throughout this file?

File: include/maidsafe/common/convert.h (hunk 2 above)
Comment (inetic) [inline include/maidsafe/common/convert.h]: I can include those, but I'm not getting errors without it, so it might be redundant.

File: src/maidsafe/common/tests/convert_test.cc [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  virtual void SetUp() { }
+  virtual void TearDown() { }
+};
+
+template<class T> void ToBoostThenBack(const T& value) {
```
Comment (Fraser999) [inline src/maidsafe/common/tests/convert_test.cc]: Again clang-format would sort this, but we normally have the `template <typename T>` on a separate line.

File: src/maidsafe/common/tests/convert_test.cc [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+template<class T> void ToAsioThenBack(const T& value) {
+  auto boost_value = common::convert::ToAsio(value);
+  auto second_value = common::convert::ToBoost(boost_value);
+  EXPECT_EQ(value, second_value);
+}
+
+TEST_F(Conversions, BEH_Address_v4) {
+  const char* ip = "123.231.213.213";
```
Comment (Fraser999) [inline src/maidsafe/common/tests/convert_test.cc]: Could you use the [`GetRandomIPv4AddressAsString()`](https://github.com/Fraser999/MaidSafe-Common/blob/next/include/maidsafe/common/test.h#L55) helper function in test.h here please so we're not using a hard-coded value?

File: src/maidsafe/common/tests/convert_test.cc [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  ToBoostThenBack(asio::ip::address_v4::from_string(ip));
+  ToAsioThenBack(boost::asio::ip::address_v4::from_string(ip));
+}
+
+TEST_F(Conversions, BEH_Address_v6) {
+  const char* ip = "2001:0db8:ac10:fe01:0000:0000:0000:0000";
```
Comment (Fraser999) [inline src/maidsafe/common/tests/convert_test.cc]: And likewise [`GetRandomIPv6AddressAsString()`](https://github.com/Fraser999/MaidSafe-Common/blob/next/include/maidsafe/common/test.h#L57) here please.

File: include/maidsafe/common/convert.h (hunk 2 above)
Comment (Fraser999) [inline include/maidsafe/common/convert.h]: Yes, but we generally aim to include what we use.  Say boost/asio/ip/udp.hpp is the only file which brings `boost::system::error_code` into scope here and we either no longer include it, or it changes to no longer include boost/system/error_code.hpp. Then that unrelated change breaks this file.

It's definitely more of an issue for headers rather than source files.

Comment (inetic) [inline include/maidsafe/common/convert.h]: I would have preferred the "fix when it becomes a problem" approach, but
changed it anyway.

File: src/maidsafe/common/tests/convert_test.cc (hunk 5 above)
Comment (inetic) [inline src/maidsafe/common/tests/convert_test.cc]: Oh, wasn't aware of the TEST macro, will change. Cheers.

File: src/maidsafe/common/tests/convert_test.cc [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  auto ip = GetRandomIPv6AddressAsString();
+  ToBoostThenBack(asio::ip::address_v6::from_string(ip));
+  ToAsioThenBack(boost::asio::ip::address_v6::from_string(ip));
+}
+
+TEST(Conversions, BEH_Address) {
+  auto ip_v4 = GetRandomIPv4AddressAsString();
+  ToBoostThenBack(asio::ip::address::from_string(ip_v4));
+  ToAsioThenBack(boost::asio::ip::address::from_string(ip_v4));
+  auto ip_v6 = GetRandomIPv6AddressAsString();
```
Comment (Fraser999) [inline src/maidsafe/common/tests/convert_test.cc]: I was just thinking of the bug I had here, and that we should maybe also set the Scope ID to a random valid value for the IPv6 test.

Comment (inetic) [inline src/maidsafe/common/tests/convert_test.cc]: All right, will add. Just curious, how did you encounter the bug? Are we testing on ipv6?

Comment (Fraser999) [inline src/maidsafe/common/tests/convert_test.cc]: I was comparing your code for converting IPv6 addresses to my code, and saw that you'd handled scope IDs where I hadn't.  I never actually saw the bug affect any of our code since I don't think we set or get scope IDs anywhere, but it was a bug nonetheless :-)
